src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt

```diff
@@ -0,0 +1,634 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.utils
+
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.lang.annotation.Annotation
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.HighlightSeverity
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElement
+import com.intellij.xml.util.XmlStringUtil.escapeString
+import org.rust.lang.core.psi.RsFunction
+import org.rust.lang.core.psi.ext.RsAbstractable
+import org.rust.lang.core.psi.ext.RsReferenceElement
+import org.rust.lang.core.types.ty.Ty
+
+sealed class RsDiagnostic(val element: PsiElement) {
+    abstract fun addToHolder(holder: AnnotationHolder): Annotation
+
+    protected fun pluralise(count: Int, singular: String, plural: String): String =
+        if (count == 1) singular else plural
+
+    class TypeError(
+        element: PsiElement,
+        private val expectedTy: Ty,
+        private val actualTy: Ty
+    ) : RsDiagnostic(element) {
+        override fun addToHolder(holder: AnnotationHolder): Annotation {
+            return holder.createErrorAnnotation(
+                element,
+                header(RsErrorCode.E0308, "mismatched types"),
+                expectedFound(expectedTy, actualTy)
+            )
+        }
+
+        private fun expectedFound(expectedTy: Ty, actualTy: Ty): String {
+            val expectedTyS = escapeString(expectedTy.toString())
+            val actualTyS = escapeString(actualTy.toString())
+            return "expected `$expectedTyS`, found `$actualTyS`"
+        }
+    }
+
+    class FieldAccessError(
```

Comment:
There is a lot of copy-paste. FieldAccessError/MethodAccessError/PathAccessError seems a very similiar. I think we can generalize it as just `AccessError` and then save `RsErrorCode` as a field (and second constructor argument) and then switch aroung it.